### PR TITLE
Instrument also directories with artifact transform 

### LINKED ISSUE #27639: Artifact transform instrumentation ignores directories
If user adds a directory to the classpath, e.g. like:
```
buildscript {
    dependencies {
        classpath(files("./included/build/classes/java/main"))
    }
}
```

then the code in the directory won't be instrumented. This is a bug introduced with artifact transform instrumentation and it worked in 8.5.

One example for such usage is testing a Gradle plugin, e.g. when usin [nebula-test plugin](https://github.com/nebula-plugins/nebula-test) with option `definePluginOutsideOfPluginBlock = true`.

This also affects transforming lambdas to serializable lambdas, you get error like:
```
* What went wrong:
Could not load the value of field `action` of `org.gradle.api.internal.AbstractTask$TaskActionWrapper` bean found in field `actions` of task `:compileJava` of type `org.gradle.api.tasks.compile.JavaCompile`.
> CustomPlugin$apply$1$$Lambda$2130/0x0000000800e80440
```

### PR DESCRIPTION
**Pull request by asodja**
Fixes #27639

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test QF

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheDirectoryPluginIntegrationTest.groovy [hunk 1]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2022 the original author or authors.
```
Comment (mlopatkin) [inline platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheDirectoryPluginIntegrationTest.groovy:2]: :nail_care: 2024?

File: subprojects/core-api/src/main/java/org/gradle/api/internal/artifacts/dsl/package-info.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright 2018 the original author or authors.
```
Comment (mlopatkin) [inline subprojects/core-api/src/main/java/org/gradle/api/internal/artifacts/dsl/package-info.java:2]: :nail_care: 2024?

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/type/DefaultArtifactTypeRegistry.java [hunk 3]
```diff
@@ -58,6 +61,7 @@ public void visitArtifactTypes(Consumer<? super ImmutableAttributes> action) {
                 if (seen.add(artifactTypeDefinition.getName())) {
                     ImmutableAttributes attributes = ((AttributeContainerInternal) artifactTypeDefinition.getAttributes()).asImmutable();
                     attributes = attributesFactory.concat(attributesFactory.of(ARTIFACT_TYPE_ATTRIBUTE, artifactTypeDefinition.getName()), attributes);
+                    attributes = attributesFactory.concat(defaultArtifactAttributes.asImmutable(), attributes);
```
Comment (mlopatkin) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/type/DefaultArtifactTypeRegistry.java:64]: :thinking: we're prepending the default attributes in many places. Maybe it is worth to try and abstract it somehow, so there's less risk of forgetting doing so on some codepath?

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/type/DefaultArtifactTypeRegistry.java
Comment (mlopatkin) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/type/DefaultArtifactTypeRegistry.java:1]: :question: Should we also have some tests for this code? We can add them as a follow-up, though, to unblock the release.

**Review by mlopatkin — APPROVED** [review]
LGTM with nits, thank you for fixing this!

File: subprojects/core-api/src/main/java/org/gradle/api/internal/artifacts/dsl/DependencyHandlerInternal.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+package org.gradle.api.internal.artifacts.dsl;
+
+import org.gradle.api.artifacts.dsl.DependencyHandler;
+import org.gradle.api.attributes.AttributeContainer;
+
+public interface DependencyHandlerInternal extends DependencyHandler {
+
+    /**
+     * Default attributes added to all artifacts.
```
Comment (jvandort) [inline subprojects/core-api/src/main/java/org/gradle/api/internal/artifacts/dsl/DependencyHandlerInternal.java:25]:
```suggestion
     * Default attributes added to all artifact variants during artifact selection
```

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/type/ArtifactTypeRegistry.java [hunk 5]
```diff
@@ -30,4 +31,6 @@ public interface ArtifactTypeRegistry extends Factory<ArtifactTypeContainer> {
     ImmutableAttributes mapAttributesFor(File file);
 
     void visitArtifactTypes(Consumer<? super ImmutableAttributes> action);
+
+    AttributeContainer getDefaultArtifactAttributes();
```
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/type/ArtifactTypeRegistry.java:35]: Can we copy-paste the javadoc from `DependencyHandlerInternal` here as well?

**Review by jvandort — APPROVED** [review]
LGTM

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/type/DefaultArtifactTypeRegistry.java
Comment (asodja) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/type/DefaultArtifactTypeRegistry.java:1]: Good point, I will add them

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheDirectoryPluginIntegrationTest.groovy (hunk 1 above)
Comment (asodja) [reply platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheDirectoryPluginIntegrationTest.groovy]: Fixed

File: subprojects/core-api/src/main/java/org/gradle/api/internal/artifacts/dsl/package-info.java (hunk 2 above)
Comment (asodja) [reply subprojects/core-api/src/main/java/org/gradle/api/internal/artifacts/dsl/package-info.java]: Fixed

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/type/DefaultArtifactTypeRegistry.java
Comment (asodja) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/type/DefaultArtifactTypeRegistry.java:1]: Added some tests

File: subprojects/core-api/src/main/java/org/gradle/api/internal/artifacts/dsl/DependencyHandlerInternal.java (hunk 4 above)
Comment (asodja) [reply subprojects/core-api/src/main/java/org/gradle/api/internal/artifacts/dsl/DependencyHandlerInternal.java]: Fixed

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/type/DefaultArtifactTypeRegistry.java (hunk 3 above)
Comment (asodja) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/type/DefaultArtifactTypeRegistry.java]: Changed code a bit so we concat default attributes in fewer places

**Comment by asodja** [comment]
@bot-gradle merge

**Pull request merged**